doc/nebula.1
.\" $Id: nebula.man,v 1.14 2020/05/21 19:31:21 lhf Exp $
.TH NEBULA 1 "$Date: 2020/05/21 19:31:21 $"
.SH NAME
nebula \- Nebula interpreter
.SH SYNOPSIS
.B nebula
[
.I options
]
[
.I script
[
.I args
]
]
.SH DESCRIPTION
.B nebula
is the standalone Nebula interpreter.
It loads and executes Nebula programs,
either in textual source form or
in precompiled binary form.
(Precompiled binaries are output by
.BR nebulac ,
the Nebula compiler.)
.B nebula
can be used as a batch interpreter and also interactively.
.LP
After handling the
.IR options ,
the Nebula program in file
.I script
is loaded and executed.
The
.I args
are available to
.I script
as strings in a global table named
.B arg
and also as arguments to its main function.
When called without arguments,
.B nebula
behaves as
.B "nebula \-v \-i"
if the standard input is a terminal,
and as
.B "nebula \-"
otherwise.
.LP
In interactive mode,
.B nebula
prompts the user,
reads lines from the standard input,
and executes them as they are read.
If the line contains an expression,
then the line is evanebulated and the result is printed.
If a line does not contain a complete statement,
then a secondary prompt is displayed and
lines are read until a complete statement is formed or
a syntax error is found.
.LP
Before handling command line options and scripts,
.B nebula
checks the contents of the environment variables
.B NEBULA_INIT_5_4
and
.BR NEBULA_INIT ,
in that order.
If the contents are of the form
.RI '@ filename ',
then
.I filename
is executed.
Otherwise, the contents are assumed to be a Nebula statement and is executed.
When
.B NEBULA_INIT_5_4
is defined,
.B NEBULA_INIT
is ignored.
.SH OPTIONS
.TP
.BI \-e " stat"
execute statement
.IR stat .
.TP
.B \-i
enter interactive mode after executing
.IR script .
.TP
.BI \-l " name"
require library
.I name
into global
.IR name .
.TP
.B \-v
show version information.
.TP
.B \-E
ignore environment variables.
.TP
.B \-W
turn warnings on.
.TP
.B \-\-
stop handling options.
.TP
.B \-
stop handling options and execute the standard input as a file.
.SH ENVIRONMENT VARIABLES
The following environment variables affect the execution of
.BR nebula .
When defined,
the version-specific variants take priority
and the version-neutral variants are ignored.
.TP
.B NEBULA_INIT, NEBULA_INIT_5_4
Code to be executed before command line options and scripts.
.TP
.B NEBULA_PATH, NEBULA_PATH_5_4
Initial value of package.cpath,
the path used by require to search for Nebula loaders.
.TP
.B NEBULA_CPATH, NEBULA_CPATH_5_4
Initial value of package.cpath,
the path used by require to search for C loaders.
.SH EXIT STATUS
If a script calls os.exit,
then
.B nebula
exits with the given exit status.
Otherwise,
.B nebula
exits
with EXIT_SUCCESS (0 on POSIX systems) if there were no errors
and
with EXIT_FAILURE (1 on POSIX systems) if there were errors.
Errors raised in interactive mode do not cause exits.
.SH DIAGNOSTICS
Error messages should be self explanatory.
.SH "SEE ALSO"
.BR nebulac (1)
.br
The documentation at nebula.org,
especially section 7 of the reference manual.
.SH AUTHORS
R. Ierusalimschy,
L. H. de Figueiredo,
W. Celes
.\" EOF


src/nebula.c
/*
** $Id: nebula.c $
** Nebula stand-alone interpreter
** See Copyright Notice in nebula.h
*/

#define nebula_c

#include "prefix.h"


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <signal.h>

#include "nebula.h"

#include "auxlib.h"
#include "nebulalib.h"


#if !defined(NEBULA_PROGNAME)
#define NEBULA_PROGNAME		"nebula"
#endif

#if !defined(NEBULA_INIT_VAR)
#define NEBULA_INIT_VAR		"NEBULA_INIT"
#endif

#define NEBULA_INITVARVERSION	NEBULA_INIT_VAR NEBULA_VERSUFFIX


static nebula_State *globalL = NULL;

static const char *progname = NEBULA_PROGNAME;


#if defined(NEBULA_USE_POSIX)   /* { */

/*
** Use 'sigaction' when available.
*/
static void setsignal (int sig, void (*handler)(int)) {
  struct sigaction sa;
  sa.sa_handler = handler;
  sa.sa_flags = 0;
  sigemptyset(&sa.sa_mask);  /* do not mask any signal */
  sigaction(sig, &sa, NULL);
}

#else           /* }{ */

#define setsignal            signal

#endif                               /* } */


/*
** Hook set by signal function to stop the interpreter.
*/
static void lstop (nebula_State *L, nebula_Debug *ar) {
  (void)ar;  /* unused arg. */
  nebula_sethook(L, NULL, 0, 0);  /* reset hook */
  nebulaL_error(L, "interrupted!");
}


/*
** Function to be called at a C signal. Because a C signal cannot
** just change a Nebula state (as there is no proper synchronization),
** this function only sets a hook that, when called, will stop the
** interpreter.
*/
static void laction (int i) {
  int flag = NEBULA_MASKCALL | NEBULA_MASKRET | NEBULA_MASKLINE | NEBULA_MASKCOUNT;
  setsignal(i, SIG_DFL); /* if another SIGINT happens, terminate process */
  nebula_sethook(globalL, lstop, flag, 1);
}


static void print_usage (const char *badoption) {
  nebula_writestringerror("%s: ", progname);
  if (badoption[1] == 'e' || badoption[1] == 'l')
    nebula_writestringerror("'%s' needs argument\n", badoption);
  else
    nebula_writestringerror("unrecognized option '%s'\n", badoption);
  nebula_writestringerror(
  "usage: %s [options] [script [args]]\n"
  "Available options are:\n"
  "  -e stat   execute string 'stat'\n"
  "  -i        enter interactive mode after executing 'script'\n"
  "  -l mod    require library 'mod' into global 'mod'\n"
  "  -l g=mod  require library 'mod' into global 'g'\n"
  "  -v        show version information\n"
  "  -E        ignore environment variables\n"
  "  -W        turn warnings on\n"
  "  --        stop handling options\n"
  "  -         stop handling options and execute stdin\n"
  ,
  progname);
}


/*
** Prints an error message, adding the program name in front of it
** (if present)
*/
static void l_message (const char *pname, const char *msg) {
  if (pname) nebula_writestringerror("%s: ", pname);
  nebula_writestringerror("%s\n", msg);
}


/*
** Check whether 'status' is not OK and, if so, prints the error
** message on the top of the stack. It assumes that the error object
** is a string, as it was either generated by Nebula or by 'msghandler'.
*/
static int report (nebula_State *L, int status) {
  if (status != NEBULA_OK) {
    const char *msg = nebula_tostring(L, -1);
    l_message(progname, msg);
    nebula_pop(L, 1);  /* remove message */
  }
  return status;
}


/*
** Message handler used to run all chunks
*/
static int msghandler (nebula_State *L) {
  const char *msg = nebula_tostring(L, 1);
  if (msg == NULL) {  /* is error object not a string? */
    if (nebulaL_callmeta(L, 1, "__tostring") &&  /* does it have a metamethod */
        nebula_type(L, -1) == NEBULA_TSTRING)  /* that produces a string? */
      return 1;  /* that is the message */
    else
      msg = nebula_pushfstring(L, "(error object is a %s value)",
                               nebulaL_typename(L, 1));
  }
  nebulaL_traceback(L, L, msg, 1);  /* append a standard traceback */
  return 1;  /* return the traceback */
}


/*
** Interface to 'nebula_pcall', which sets appropriate message function
** and C-signal handler. Used to run all chunks.
*/
static int docall (nebula_State *L, int narg, int nres) {
  int status;
  int base = nebula_gettop(L) - narg;  /* function index */
  nebula_pushcfunction(L, msghandler);  /* push message handler */
  nebula_insert(L, base);  /* put it under function and args */
  globalL = L;  /* to be available to 'laction' */
  setsignal(SIGINT, laction);  /* set C-signal handler */
  status = nebula_pcall(L, narg, nres, base);
  setsignal(SIGINT, SIG_DFL); /* reset C-signal handler */
  nebula_remove(L, base);  /* remove message handler from the stack */
  return status;
}


static void print_version (void) {
  nebula_writestring(NEBULA_COPYRIGHT, strlen(NEBULA_COPYRIGHT));
  nebula_writeline();
}


/*
** Create the 'arg' table, which stores all arguments from the
** command line ('argv'). It should be aligned so that, at index 0,
** it has 'argv[script]', which is the script name. The arguments
** to the script (everything after 'script') Nebula to positive indices;
** other arguments (before the script name) Nebula to negative indices.
** If there is no script name, assume interpreter's name as base.
*/
static void createargtable (nebula_State *L, char **argv, int argc, int script) {
  int i, narg;
  if (script == argc) script = 0;  /* no script name? */
  narg = argc - (script + 1);  /* number of positive indices */
  nebula_createtable(L, narg, script + 1);
  for (i = 0; i < argc; i++) {
    nebula_pushstring(L, argv[i]);
    nebula_rawseti(L, -2, i - script);
  }
  nebula_setglobal(L, "arg");
}


static int dochunk (nebula_State *L, int status) {
  if (status == NEBULA_OK) status = docall(L, 0, 0);
  return report(L, status);
}


static int dofile (nebula_State *L, const char *name) {
  return dochunk(L, nebulaL_loadfile(L, name));
}


static int dostring (nebula_State *L, const char *s, const char *name) {
  return dochunk(L, nebulaL_loadbuffer(L, s, strlen(s), name));
}


/*
** Receives 'globname[=modname]' and runs 'globname = require(modname)'.
*/
static int dolibrary (nebula_State *L, char *globname) {
  int status;
  char *modname = strchr(globname, '=');
  if (modname == NULL)  /* no explicit name? */
    modname = globname;  /* module name is equal to global name */
  else {
    *modname = '\0';  /* global name ends here */
    modname++;  /* module name starts after the '=' */
  }
  nebula_getglobal(L, "require");
  nebula_pushstring(L, modname);
  status = docall(L, 1, 1);  /* call 'require(modname)' */
  if (status == NEBULA_OK)
    nebula_setglobal(L, globname);  /* globname = require(modname) */
  return report(L, status);
}


/*
** Push on the stack the contents of table 'arg' from 1 to #arg
*/
static int pushargs (nebula_State *L) {
  int i, n;
  if (nebula_getglobal(L, "arg") != NEBULA_TTABLE)
    nebulaL_error(L, "'arg' is not a table");
  n = (int)nebulaL_len(L, -1);
  nebulaL_checkstack(L, n + 3, "too many arguments to script");
  for (i = 1; i <= n; i++)
    nebula_rawgeti(L, -i, i);
  nebula_remove(L, -i);  /* remove table from the stack */
  return n;
}


static int handle_script (nebula_State *L, char **argv) {
  int status;
  const char *fname = argv[0];
  if (strcmp(fname, "-") == 0 && strcmp(argv[-1], "--") != 0)
    fname = NULL;  /* stdin */
  status = nebulaL_loadfile(L, fname);
  if (status == NEBULA_OK) {
    int n = pushargs(L);  /* push arguments to script */
    status = docall(L, n, NEBULA_MULTRET);
  }
  return report(L, status);
}


/* bits of various argument indicators in 'args' */
#define has_error	1	/* bad option */
#define has_i		2	/* -i */
#define has_v		4	/* -v */
#define has_e		8	/* -e */
#define has_E		16	/* -E */


/*
** Traverses all arguments from 'argv', returning a mask with those
** needed before running any Nebula code (or an error code if it finds
** any invalid argument). 'first' returns the first not-handled argument
** (either the script name or a bad argument in case of error).
*/
static int collectargs (char **argv, int *first) {
  int args = 0;
  int i;
  for (i = 1; argv[i] != NULL; i++) {
    *first = i;
    if (argv[i][0] != '-')  /* not an option? */
        return args;  /* stop handling options */
    switch (argv[i][1]) {  /* else check option */
      case '-':  /* '--' */
        if (argv[i][2] != '\0')  /* extra characters after '--'? */
          return has_error;  /* invalid option */
        *first = i + 1;
        return args;
      case '\0':  /* '-' */
        return args;  /* script "name" is '-' */
      case 'E':
        if (argv[i][2] != '\0')  /* extra characters? */
          return has_error;  /* invalid option */
        args |= has_E;
        break;
      case 'W':
        if (argv[i][2] != '\0')  /* extra characters? */
          return has_error;  /* invalid option */
        break;
      case 'i':
        args |= has_i;  /* (-i implies -v) *//* FALLTHROUGH */
      case 'v':
        if (argv[i][2] != '\0')  /* extra characters? */
          return has_error;  /* invalid option */
        args |= has_v;
        break;
      case 'e':
        args |= has_e;  /* FALLTHROUGH */
      case 'l':  /* both options need an argument */
        if (argv[i][2] == '\0') {  /* no concatenated argument? */
          i++;  /* try next 'argv' */
          if (argv[i] == NULL || argv[i][0] == '-')
            return has_error;  /* no next argument or it is another option */
        }
        break;
      default:  /* invalid option */
        return has_error;
    }
  }
  *first = i;  /* no script name */
  return args;
}


/*
** Processes options 'e' and 'l', which involve running Nebula code, and
** 'W', which also affects the state.
** Returns 0 if some code raises an error.
*/
static int runargs (nebula_State *L, char **argv, int n) {
  int i;
  for (i = 1; i < n; i++) {
    int option = argv[i][1];
    nebula_assert(argv[i][0] == '-');  /* already checked */
    switch (option) {
      case 'e':  case 'l': {
        int status;
        char *extra = argv[i] + 2;  /* both options need an argument */
        if (*extra == '\0') extra = argv[++i];
        nebula_assert(extra != NULL);
        status = (option == 'e')
                 ? dostring(L, extra, "=(command line)")
                 : dolibrary(L, extra);
        if (status != NEBULA_OK) return 0;
        break;
      }
      case 'W':
        nebula_warning(L, "@on", 0);  /* warnings on */
        break;
    }
  }
  return 1;
}


static int handle_nebulainit (nebula_State *L) {
  const char *name = "=" NEBULA_INITVARVERSION;
  const char *init = getenv(name + 1);
  if (init == NULL) {
    name = "=" NEBULA_INIT_VAR;
    init = getenv(name + 1);  /* try alternative name */
  }
  if (init == NULL) return NEBULA_OK;
  else if (init[0] == '@')
    return dofile(L, init+1);
  else
    return dostring(L, init, name);
}


/*
** {==================================================================
** Read-Eval-Print Loop (REPL)
** ===================================================================
*/

#if !defined(NEBULA_PROMPT)
#define NEBULA_PROMPT		"\x1b[35m⭐ ==> \x1b[0m"
#define NEBULA_PROMPT2		"\x1b[35m💫 ==>> \x1b[0m"
#endif

#if !defined(NEBULA_MAXINPUT)
#define NEBULA_MAXINPUT		512
#endif

/*
** nebula_stdin_is_tty detects whether the standard input is a 'tty' (that
** is, whether we're running nebula interactively).
*/
#if !defined(nebula_stdin_is_tty)	/* { */

#if defined(NEBULA_USE_POSIX)	/* { */

#include <unistd.h>
#define nebula_stdin_is_tty()	isatty(0)

#elif defined(NEBULA_USE_WINDOWS)	/* }{ */

#include <io.h>
#include <windows.h>

#define nebula_stdin_is_tty()	_isatty(_fileno(stdin))

#else				/* }{ */

/* ISO C definition */
#define nebula_stdin_is_tty()	1  /* assume stdin is a tty */

#endif				/* } */

#endif				/* } */


/*
** nebula_readline defines how to show a prompt and then read a line from
** the standard input.
** nebula_saveline defines how to "save" a read line in a "history".
** nebula_freeline defines how to free a line read by nebula_readline.
*/
#if !defined(nebula_readline)	/* { */

#if defined(NEBULA_USE_READLINE)	/* { */

#include <readline/readline.h>
#include <readline/history.h>
#define nebula_initreadline(L)	((void)L, rl_readline_name="nebula")
#define nebula_readline(L,b,p)	((void)L, ((b)=readline(p)) != NULL)
#define nebula_saveline(L,line)	((void)L, add_history(line))
#define nebula_freeline(L,b)	((void)L, free(b))

#else				/* }{ */

#define nebula_initreadline(L)  ((void)L)
#define nebula_readline(L,b,p) \
        ((void)L, fputs(p, stdout), fflush(stdout),  /* show prompt */ \
        fgets(b, NEBULA_MAXINPUT, stdin) != NULL)  /* get line */
#define nebula_saveline(L,line)	{ (void)L; (void)line; }
#define nebula_freeline(L,b)	{ (void)L; (void)b; }

#endif				/* } */

#endif				/* } */


/*
** Return the string to be used as a prompt by the interpreter. Leave
** the string (or nil, if using the default value) on the stack, to keep
** it anchored.
*/
static const char *get_prompt (nebula_State *L, int firstline) {
  if (nebula_getglobal(L, firstline ? "_PROMPT" : "_PROMPT2") == NEBULA_TNIL)
    return (firstline ? NEBULA_PROMPT : NEBULA_PROMPT2);  /* use the default */
  else {  /* apply 'tostring' over the value */
    const char *p = nebulaL_tolstring(L, -1, NULL);
    nebula_remove(L, -2);  /* remove original value */
    return p;
  }
}

/* mark in error messages for incomplete statements */
#define EOFMARK		"<eof>"
#define marklen		(sizeof(EOFMARK)/sizeof(char) - 1)


/*
** Check whether 'status' signals a syntax error and the error
** message at the top of the stack ends with the above mark for
** incomplete statements.
*/
static int incomplete (nebula_State *L, int status) {
  if (status == NEBULA_ERRSYNTAX) {
    size_t lmsg;
    const char *msg = nebula_tolstring(L, -1, &lmsg);
    if (lmsg >= marklen && strcmp(msg + lmsg - marklen, EOFMARK) == 0) {
      nebula_pop(L, 1);
      return 1;
    }
  }
  return 0;  /* else... */
}


/*
** Prompt the user, read a line, and push it into the Nebula stack.
*/
static int pushline (nebula_State *L, int firstline) {
  char buffer[NEBULA_MAXINPUT];
  char *b = buffer;
  size_t l;
  const char *prmt = get_prompt(L, firstline);
  int readstatus = nebula_readline(L, b, prmt);
  if (readstatus == 0)
    return 0;  /* no input (prompt will be popped by caller) */
  nebula_pop(L, 1);  /* remove prompt */
  l = strlen(b);
  if (l > 0 && b[l-1] == '\n')  /* line ends with newline? */
    b[--l] = '\0';  /* remove it */
  if (firstline && b[0] == '=')  /* for compatibility with 5.2, ... */
    nebula_pushfstring(L, "return %s", b + 1);  /* change '=' to 'return' */
  else
    nebula_pushlstring(L, b, l);
  nebula_freeline(L, b);
  return 1;
}


/*
** Try to compile line on the stack as 'return <line>;'; on return, stack
** has either compiled chunk or original line (if compilation failed).
*/
static int addreturn (nebula_State *L) {
  const char *line = nebula_tostring(L, -1);  /* original line */
  const char *retline = nebula_pushfstring(L, "return %s;", line);
  int status = nebulaL_loadbuffer(L, retline, strlen(retline), "=stdin");
  if (status == NEBULA_OK) {
    nebula_remove(L, -2);  /* remove modified line */
    if (line[0] != '\0')  /* non empty? */
      nebula_saveline(L, line);  /* keep history */
  }
  else
    nebula_pop(L, 2);  /* pop result from 'nebulaL_loadbuffer' and modified line */
  return status;
}


/*
** Read multiple lines until a complete Nebula statement
*/
static int multiline (nebula_State *L) {
  for (;;) {  /* repeat until gets a complete statement */
    size_t len;
    const char *line = nebula_tolstring(L, 1, &len);  /* get what it has */
    int status = nebulaL_loadbuffer(L, line, len, "=stdin");  /* try it */
    if (!incomplete(L, status) || !pushline(L, 0)) {
      nebula_saveline(L, line);  /* keep history */
      return status;  /* cannot or should not try to add continuation line */
    }
    nebula_pushliteral(L, "\n");  /* add newline... */
    nebula_insert(L, -2);  /* ...between the two lines */
    nebula_concat(L, 3);  /* join them */
  }
}


/*
** Read a line and try to load (compile) it first as an expression (by
** adding "return " in front of it) and second as a statement. Return
** the final status of load/call with the resulting function (if any)
** in the top of the stack.
*/
static int loadline (nebula_State *L) {
  int status;
  nebula_settop(L, 0);
  if (!pushline(L, 1))
    return -1;  /* no input */
  if ((status = addreturn(L)) != NEBULA_OK)  /* 'return ...' did not work? */
    status = multiline(L);  /* try as command, maybe with continuation lines */
  nebula_remove(L, 1);  /* remove line from the stack */
  nebula_assert(nebula_gettop(L) == 1);
  return status;
}


/*
** Prints (calling the Nebula 'print' function) any values on the stack
*/
static void l_print (nebula_State *L) {
  int n = nebula_gettop(L);
  if (n > 0) {  /* any result to be printed? */
    nebulaL_checkstack(L, NEBULA_MINSTACK, "too many results to print");
    nebula_getglobal(L, "print");
    nebula_insert(L, 1);
    if (nebula_pcall(L, n, 0, 0) != NEBULA_OK)
      l_message(progname, nebula_pushfstring(L, "error calling 'print' (%s)",
                                             nebula_tostring(L, -1)));
  }
}


/*
** Do the REPL: repeatedly read (load) a line, evanebulate (call) it, and
** print any results.
*/
static void doREPL (nebula_State *L) {
  int status;
  const char *oldprogname = progname;
  progname = NULL;  /* no 'progname' on errors in interactive mode */
  nebula_initreadline(L);
  while ((status = loadline(L)) != -1) {
    if (status == NEBULA_OK)
      status = docall(L, 0, NEBULA_MULTRET);
    if (status == NEBULA_OK) l_print(L);
    else report(L, status);
  }
  nebula_settop(L, 0);  /* clear stack */
  nebula_writeline();
  progname = oldprogname;
}

/* }================================================================== */


/*
** Main body of stand-alone interpreter (to be called in protected mode).
** Reads the options and handles them all.
*/
static int pmain (nebula_State *L) {
  int argc = (int)nebula_tointeger(L, 1);
  char **argv = (char **)nebula_touserdata(L, 2);
  int script;
  int args = collectargs(argv, &script);
  nebulaL_checkversion(L);  /* check that interpreter has correct version */
  if (argv[0] && argv[0][0]) progname = argv[0];
  if (args == has_error) {  /* bad arg? */
    print_usage(argv[script]);  /* 'script' has index of bad arg. */
    return 0;
  }
  if (args & has_v)  /* option '-v'? */
    print_version();
  if (args & has_E) {  /* option '-E'? */
    nebula_pushboolean(L, 1);  /* signal for libraries to ignore env. vars. */
    nebula_setfield(L, NEBULA_REGISTRYINDEX, "NEBULA_NOENV");
  }
  nebulaL_openlibs(L);  /* open standard libraries */
  createargtable(L, argv, argc, script);  /* create table 'arg' */
  nebula_gc(L, NEBULA_GCGEN, 0, 0);  /* GC in generational mode */
  if (!(args & has_E)) {  /* no option '-E'? */
    if (handle_nebulainit(L) != NEBULA_OK)  /* run NEBULA_INIT */
      return 0;  /* error running NEBULA_INIT */
  }
  if (!runargs(L, argv, script))  /* execute arguments -e and -l */
    return 0;  /* something failed */
  if (script < argc &&  /* execute main script (if there is one) */
      handle_script(L, argv + script) != NEBULA_OK)
    return 0;
  if (args & has_i)  /* -i option? */
    doREPL(L);  /* do read-eval-print loop */
  else if (script == argc && !(args & (has_e | has_v))) {  /* no arguments? */
    if (nebula_stdin_is_tty()) {  /* running in interactive mode? */
      print_version();
      doREPL(L);  /* do read-eval-print loop */
    }
    else dofile(L, NULL);  /* executes stdin as a file */
  }
  nebula_pushboolean(L, 1);  /* signal no errors */
  return 1;
}


int main (int argc, char **argv) {
  int status, result;
  nebula_State *L = nebulaL_newstate();  /* create state */
  if (L == NULL) {
    l_message(argv[0], "cannot create state: not enough memory");
    return EXIT_FAILURE;
  }
  nebula_pushcfunction(L, &pmain);  /* to call 'pmain' in protected mode */
  nebula_pushinteger(L, argc);  /* 1st argument */
  nebula_pushlightuserdata(L, argv); /* 2nd argument */
  status = nebula_pcall(L, 2, 1, 0);  /* do the call */
  result = nebula_toboolean(L, -1);  /* get result */
  report(L, status);
  nebula_close(L);
  return (result && status == NEBULA_OK) ? EXIT_SUCCESS : EXIT_FAILURE;
}